### PR TITLE
expression: cast charset according to the function's resulting charset

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: 
close https://github.com/pingcap/tidb/issues/23506
close https://github.com/pingcap/tidb/issues/29608
close https://github.com/pingcap/tidb/issues/26818

we add a `from_binary` function and make it can be pushdown to tikv. see the related `to_binary` https://github.com/pingcap/tidb/pull/29736

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/builtin_convert_charset.go [hunk 1]
```diff
@@ -108,28 +127,155 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
...
+		RetType:  tp,
+		Function: f,
+	}
+	return FoldConstant(res)
+}
+
+// HandleBinaryLiteral wraps `expr` with to_binary or from_binary sig.
+func HandleBinaryLiteral(ctx sessionctx.Context, expr Expression, ec *ExprCollation, funcName string) Expression {
+	switch funcName {
+	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Lcase, ast.Reverse, ast.Upper, ast.Ucase, ast.Quote, ast.Coalesce,
```
Comment (AilinKid) [inline expression/builtin_convert_charset.go:264]: Rest looks good to me, this part set I will left it for @tangenta to take it

File: expression/builtin_convert_charset.go [hunk 2]
```diff
@@ -108,28 +127,155 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
...
+		ast.FindInSet, ast.Regexp, ast.Field, ast.Locate, ast.Instr, ast.Position, ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ,
+		ast.NE, ast.NullEQ, ast.Strcmp, ast.If, ast.Ifnull, ast.Like, ast.In, ast.DateFormat, ast.TimeFormat:
+		if ec.Charset == charset.CharsetBin && expr.GetType().Charset != charset.CharsetBin {
+			return BuildToBinaryFunction(ctx, expr)
+		} else if ec.Charset != charset.CharsetBin && expr.GetType().Charset == charset.CharsetBin {
+			ft := expr.GetType().Clone()
+			ft.Charset, ft.Collate = ec.Charset, ec.Collation
+			return BuildFromBinaryFunction(ctx, expr, ft)
+		}
+	case ast.Hex, ast.Length, ast.OctetLength, ast.ASCII, ast.ToBase64:
```
Comment (Defined2014) [inline expression/builtin_convert_charset.go:276]: We can add `ast.MD5`, `ast.SHA1` and else here, then delete related code in expression.

Comment (xiongjiwei) [reply expression/builtin_convert_charset.go:276]: yes, and it will be in another pr

Comment (Defined2014) [reply expression/builtin_convert_charset.go:276]: Found it, #29950

File: expression/builtin_convert_charset.go [hunk 3]
```diff
@@ -108,28 +127,154 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
...
+		return err
+	}
+	transferString := b.getTransferFunc()
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		str, err := transferString(buf.GetBytes(i))
```
Comment (tangenta) [inline expression/builtin_convert_charset.go:198]: We can use a buffer to reduce memory allocation.

Comment (xiongjiwei) [reply expression/builtin_convert_charset.go:198]: yes, it has the buffer, see https://github.com/pingcap/tidb/pull/29905/files#diff-b7c01f9c9ab46f762c72934562d5e01b7ec3750c1b6b6601147b5b2808d30d5cR218

File: cmd/explaintest/r/new_character_set_builtin.result [hunk 4]
```diff
@@ -286,3 +286,207 @@ select ord(a), ord(b), ord(c) from t;
...
+E4BA8CE4BA8CE4BA8CE4BA8CE4B880	B6FEB6FEB6FEB6FED2BB
+E4BA8CE4BA8CE4BA8CE4BA8CE4B880	B6FEB6FEB6FEB6FED2BB
+E4BA8CE4BA8CE4BA8CE4BA8CE4B880	B6FEB6FEB6FEB6FED2BB
+select hex(lpad(a, 5, c)), hex(lpad(b, 5, c)) from t;
+hex(lpad(a, 5, c))	hex(lpad(b, 5, c))
+E4B8E4B880	E4B880D2BB
+D2BBE4B880	D2BB00D2BB
+E4BAE4B880	E4BA8CD2BB
+B6FEE4B880	B6FE00D2BB
+select hex(rpad(a, 5, 0xe4ba8c)), hex(rpad(b, 5, 0xb6fe)) from t;
```
Comment (xiongjiwei) [inline cmd/explaintest/r/new_character_set_builtin.result:344]: mysql has different results here and they confirmed it is bug https://bugs.mysql.com/bug.php?id=105668&thanks=4

**Review by tangenta — APPROVED** [review]

File: expression/builtin_convert_charset.go [hunk 5]
```diff
@@ -108,28 +127,154 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
...
+				return nil, errCannotConvertString.GenWithStackByArgs(fmt.Sprintf("%X", s), charset.CharsetBin, b.tp.Charset)
 			}
-			return FoldConstant(sf)
+			return str, nil
+		}
+	}
+	return transferString
+}
+
+// BuildToBinaryFunction to_binary function.
```
Comment (zimulala) [inline expression/builtin_convert_charset.go:230]:
```suggestion
// BuildToBinaryFunction builds to_binary function.
```

File: expression/builtin_convert_charset.go [hunk 6]
```diff
@@ -45,6 +64,8 @@ func (c *tidbConvertCharsetFunctionClass) getFunction(ctx sessionctx.Context, ar
 		if err != nil {
 			return nil, err
 		}
+		bf.tp = args[0].GetType().Clone()
+		bf.tp.Charset, bf.tp.Collate = charset.CharsetBin, charset.CollationBin
```
Comment (zimulala) [inline expression/builtin_convert_charset.go:68]: It seems that we don't use `bf.tp`, we use `b.args[0].GetType` directly: https://github.com/pingcap/tidb/pull/29905/files#diff-b7c01f9c9ab46f762c72934562d5e01b7ec3750c1b6b6601147b5b2808d30d5cR92.

File: expression/collation.go [hunk 7]
```diff
@@ -301,7 +301,8 @@ func safeConvert(ctx sessionctx.Context, ec *ExprCollation, args ...Expression)
 			continue
 		}
 
-		if arg.Repertoire() == ASCII {
+		// if value has ASCII repertoire, or it is binary string, just skip it.
```
Comment (zimulala) [inline expression/collation.go:304]:
```suggestion
		// If value has ASCII repertoire, or it is binary string, just skip it.
```

File: expression/constant_test.go [hunk 8]
```diff
@@ -239,16 +238,31 @@ func TestConstantFoldingCharsetConvert(t *testing.T) {
...
+				InternalFuncFromBinary, types.NewFieldType(mysql.TypeVarchar),
+				newString("中文", "binary"))),
+			result: "中文",
+		},
+		{
+			condition: newFunction(ast.Concat,
+				newFunctionWithType(
+					InternalFuncFromBinary, types.NewFieldTypeWithCollation(mysql.TypeVarchar, "gbk_bin", -1),
+					newString("\xd2\xbb", "binary")),
+				newString("中文", "gbk_bin"),
```
Comment (zimulala) [inline expression/constant_test.go:262]: How about changing the two args place, like `newString("中文", "gbk_bin"), newString("\xd2\xbb", "binary"))`?

File: cmd/explaintest/r/new_character_set_builtin.result (hunk 4 above)
Comment (zimulala) [reply cmd/explaintest/r/new_character_set_builtin.result:344]: Please make a note of it and put it in our discussion document.

File: expression/builtin_convert_charset.go (hunk 6 above)
Comment (xiongjiwei) [reply expression/builtin_convert_charset.go:68]: yes, it is not used for `evalString`, it is set the function's return type

File: expression/constant_test.go (hunk 8 above)
Comment (xiongjiwei) [reply expression/constant_test.go:260]: updated

File: expression/builtin_convert_charset.go (hunk 6 above)
Comment (zimulala) [reply expression/builtin_convert_charset.go:68]: Got it.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**